pending-documentation.md

```diff
@@ -1665,6 +1665,72 @@ Since this feature is not clearly needed by K users at the moment, it is only
 presented here as an example; its implementation will be postponed for such time
 when its usefulness becomes apparent.
 
+Parser Generation
+-----------------
+
+In addition to on-the-fly parser generation using `kast`, K is capable of
+ahead-of-time parser generation of LR(1) or GLR parsers using Flex and Bison.
+This can be done one of two different ways.
+
+1. You can explicitly request for a particular parser to be generated by
+   invoking `kast --gen-parser <outputFile>` or
+   `kast --gen-glr-parser <outputFile>` respectively. `kast` will then create a
+   parser based on the same command line flags that govern on-the-fly parsing,
+   like `-s` to specify the starting sort, and `-m` to specify the module to
+   parse under. By default, this generates a parser for the sort of the $PGM
+   configuration variable in the main syntax module of the definition.
+2. You can request that a specific set of parsers be generated for all the
+   configuration variables of your definition by passing the
+   `--gen-bison-parser` or `--gen-glr-bison-parser` flags to `kompile`.
+   `kompile` will decide the sorts to use as start symbols based on the sorts
+   in the configuration declaration for the configuration variables. The $PGM
+   configuration variable will be generated based on the main syntax module
+   of the definition. The user must explicitly annotate other modules with the
+   configuration variables that they should be used to parse as attributes.
+   For example, `module TEST [parser(FOO)]` will tell `kompile` to parse the 
+   $FOO configuration variable using the syntax in the TEST module. If the user
+   forgets to annotate the modules with the configuration variables, only the 
+   $PGM parser will be generated.
```

Comment:
I don't particularly like this way of specifying the module to parse in, because it isn't local to where the variable is declared. For example, what if you want to use the `INT-SYNTAX` module to parse a given configuration variable?

I think it's better to either,

1.   have the annotation on the cell itself (eg. `<foo parseModule="FOO-SYNTAX"> $FOO:Foo </foo>)
2.   as an attribute on the module which declares the configuration
3.   don't directly generate parsers for configuration variables, instead for pairs of `(Sort, Module)`. So we would generate `parser_Pgm_MAIN-SYNTAX-MODULE`, `parser_Pgm_SOME-OTHER-MODULE`, `parser_Foo_FOO-SYNTAX`, `parser_Foo_MAIN-SYNTAX-MODULE`, where `Pgm` and `Foo` are sorts, not configuration variables. We can still always generate the default `parser_Pgm_MAIN-SYNTAX-MODULE`. Then `kompile/kast` can take a variadic number of Sort/Module pairs (default just being `[(sort(PGM), MAIN-SYNTAX-MODULE)]`) to generate the parsers.

In order of preference, I would say (3), (1), (2). Mostly I like (3) because it makes it exposes what the actual functionality is, which has little to do with the configuration variables, and more to do with parsers for sorts in modules. The only reason I see it getting conflated with configuration variables is because those are the most useful parsers to generate.